# Pandas
* Pandas is a open source python library designed specifically for accomplishing data analysis tasks.

In [2]:
import numpy as np
import pandas as pd

In [3]:
data = [10, 20, 30, 40, 50]
type(data)

list

In [4]:
series = pd.Series(data)
print(series)

0    10
1    20
2    30
3    40
4    50
dtype: int64


In [5]:
type(series)

pandas.core.series.Series

In [6]:
dict = {
    "name": "Rajendra",
    "age": 23,
    "address": "Bhaktapur"
}

In [7]:
dict_series = pd.Series(dict)

In [8]:
type(dict_series)

pandas.core.series.Series

In [9]:
arr = np.array(data)
arr

array([10, 20, 30, 40, 50])

In [10]:
arr_series = pd.Series(arr)
arr_series

0    10
1    20
2    30
3    40
4    50
dtype: int32

In [11]:
type(arr_series)

pandas.core.series.Series

In [12]:
arr_series.shape

(5,)

# DataFrame

In [13]:
nested_list = [[30, 40, 20], [20, 50, 10], [20, 50, 20]]

In [14]:
pd.DataFrame(nested_list, columns=["col1", "col2", "col3"])

col1  col2  col3
0    30    40    20
1    20    50    10
2    20    50    20

In [15]:
nested_list[1][-1]

10

In [16]:
len(nested_list)

3

In [17]:
data_dict = {
    "Name": ["Rajendra", "Naresh", "Mohan"],
    "Age": [23, 23, 34]
}

In [18]:
dict_df = pd.DataFrame(data_dict)

In [19]:
dict_df

Name  Age
0  Rajendra   23
1    Naresh   23
2     Mohan   34

In [20]:
dict_df.columns = ["name", "age"]

In [21]:
dict_df

name  age
0  Rajendra   23
1    Naresh   23
2     Mohan   34

In [22]:
type(dict_df)

pandas.core.frame.DataFrame

In [23]:
list_dict = [{"name": "Rajendra", "age": 23, "address": "Bhaktapur"}, {"name": "Mohan", "age": 33, "address": "panauti"}, {"name": "Naresh", "address": "Bhaktapur"}]

In [24]:
pd.DataFrame(list_dict)

name   age    address
0  Rajendra  23.0  Bhaktapur
1     Mohan  33.0    panauti
2    Naresh   NaN  Bhaktapur

In [36]:
df = pd.read_csv("car_details.csv")

In [37]:
df.head()

name  year  selling_price  km_driven    fuel  \
0             Maruti 800 AC  2007          60000      70000  Petrol   
1  Maruti Wagon R LXI Minor  2007         135000      50000  Petrol   
2      Hyundai Verna 1.6 SX  2012         600000     100000  Diesel   
3    Datsun RediGO T Option  2017         250000      46000  Petrol   
4     Honda Amaze VX i-DTEC  2014         450000     141000  Diesel   

  seller_type transmission         owner  
0  Individual       Manual   First Owner  
1  Individual       Manual   First Owner  
2  Individual       Manual   First Owner  
3  Individual       Manual   First Owner  
4  Individual       Manual  Second Owner

In [28]:
# import openpyxl
# df.to_excel("excel.xlsx", "Sheet1")

In [88]:
df_hr = pd.read_csv("data/HR_comma_sep.csv")

In [89]:
df_hr.head() # first five rows

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [90]:
df_hr.tail() # last five rows

satisfaction_level  last_evaluation  number_project  \
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department salary  
14994                      0    support    low  
14995                      0    support    low  
14996                      0    support    low  
14997                      0    support    low  
14998                      0    support    low

In [91]:
df_hr.shape # shape of DataFrame

(14999, 10)

In [92]:
df_hr.index # gives total index and step

RangeIndex(start=0, stop=14999, step=1)

In [93]:
nested_list = [[4, 2, 5], [2, 5, 2]]
pd.DataFrame(nested_list)

0  1  2
0  4  2  5
1  2  5  2

In [94]:
df_hr.columns # gives columns name in list

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [95]:
df_hr.dtypes # gives overall datatypes columns data

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

In [96]:
hr = df_hr

In [97]:
hr.rename(columns={"last_evaluation": "le"}, inplace=True)

In [98]:
hr.head(2)

satisfaction_level    le  number_project  average_montly_hours  \
0                0.38  0.53               2                   157   
1                0.80  0.86               5                   262   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   

   salary  
0     low  
1  medium

In [99]:
df_hr.head(2)

satisfaction_level    le  number_project  average_montly_hours  \
0                0.38  0.53               2                   157   
1                0.80  0.86               5                   262   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   

   salary  
0     low  
1  medium

In [106]:
df_hr.to_numpy() # convert the data to numpy array

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [107]:
df_hr.info() # gives class, index, columns, datatypes, null, not-null

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   le                     14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


* DataFrame occupies 1.1+ MB memory
* Columns are of types: float64, int64, object
* Total rows = 14999
* No any columns contains null values

In [108]:
df_hr.values # same as .to_numpy()

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [109]:
df_hr.describe()

satisfaction_level            le  number_project  average_montly_hours  \
count        14999.000000  14999.000000    14999.000000          14999.000000   
mean             0.612834      0.716102        3.803054            201.050337   
std              0.248631      0.171169        1.232592             49.943099   
min              0.090000      0.360000        2.000000             96.000000   
25%              0.440000      0.560000        3.000000            156.000000   
50%              0.640000      0.720000        4.000000            200.000000   
75%              0.820000      0.870000        5.000000            245.000000   
max              1.000000      1.000000        7.000000            310.000000   

       time_spend_company  Work_accident          left  promotion_last_5years  
count        14999.000000   14999.000000  14999.000000           14999.000000  
mean             3.498233       0.144610      0.238083               0.021268  
std              1.460136       0.351719      0.425924               0.144281  
min              2.000000       0.000000      0.000000               0.000000  
25%              3.000000       0.000000      0.000000               0.000000  
50%              3.000000       0.000000      0.000000               0.000000  
75%              4.000000       0.000000      0.000000               0.000000  
max             10.000000       1.000000      1.000000               1.000000

* We get the summary of each DataFrame columns easily without the need for explicit computation
* By default it includes only the numerical columns
* Lets get description of object type only in below cells

In [112]:
df_hr.describe(include="object")

Department salary
count       14999  14999
unique         10      3
top         sales    low
freq         4140   7316

In [114]:
df_hr.describe(exclude=["int", "float"])

Department salary
count       14999  14999
unique         10      3
top         sales    low
freq         4140   7316

* For Object type, we get count, unique, top items with frequency.

# Data Selection
* This section covers different techniques to access columns, slice DataFrame, get unique columns values:
  * select columns (using[], and dot(.) operator)
  * iloc
  * loc
  * unique() and nunique()
  * value_counts()

**select columns**

In [115]:
print(df_hr["satisfaction_level"])

print("---------------------")

0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64
---------------------


In [116]:
df_hr.satisfaction_level

0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64

# multiple columns

In [127]:
columns_to_select = ["satisfaction_level", "left"]

In [125]:
df_hr[columns_to_select]

satisfaction_level  left
0                    0.38     1
1                    0.80     1
2                    0.11     1
3                    0.72     1
4                    0.37     1
...                   ...   ...
14994                0.40     1
14995                0.37     1
14996                0.37     1
14997                0.11     1
14998                0.37     1

[14999 rows x 2 columns]

# iloc
* ways to slice pandas DataFrame via integer value
* it allows you to select data fram a DataFrame using integer indices for both rows and columns
* 1st rows is 0, 2nd row is 1, and so on
* 1st columns is 0, 2,d columns is 1 and so on.
* 

In [132]:
df_hr.iloc[0, 1]

0.53

In [133]:
df_hr.iloc[:3, -3:]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

In [134]:
df_hr.head(3)

satisfaction_level    le  number_project  average_montly_hours  \
0                0.38  0.53               2                   157   
1                0.80  0.86               5                   262   
2                0.11  0.88               7                   272   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium

# loc
* loc is used for label-based indexing
* Allows us to slice DF using row label and colums label instead of integer values as in loc
* syntax:
  ```python
  df.loc[row_label, column_label]
  ```

In [135]:
df_hr.loc[0, "satisfaction_level"]

0.38

In [136]:
row_label = [0, 1, 2]
column_label = ["promotion_last_5years", "Department", "salary"]
df_hr.loc[row_label, column_label]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

# unique() and nunique()
* unique() => gives unique values in columns in array
* nunique() => gives count of unique values

In [137]:
df_hr["left"].unique() 

array([1, 0], dtype=int64)

In [141]:
# This means left columns have 2 unique values 0 and 1

In [139]:
df_hr["salary"].unique()

array(['low', 'medium', 'high'], dtype=object)

In [140]:
df_hr["salary"].nunique()

3

In [142]:
df_hr["left"].value_counts()

left
0    11428
1     3571
Name: count, dtype: int64

In [143]:
df_hr["salary"].value_counts()

salary
low       7316
medium    6446
high      1237
Name: count, dtype: int64

# Aggregations
* Methods are:
  * mean()
  * std()
  * sum()
  * max()
  * min()

In [166]:
num_hr = df_hr["satisfaction_level"].mean()
# num_hr.mean(axis=0)

In [168]:
cat_hr = df_hr.salary
cat_hr.mode()

0    low
Name: salary, dtype: object

In [165]:
df_hr["satisfaction_level"].sum()

9191.89

# Missing Data
* Finding and Handling missing data
* methods are
  * isnull() or isna()
  * notnull() or notna()
  * dropna()
  * fillna()

# isnull() or isna()
* returns True if DataFrame values are null
* DataFrame method to get NAN values from columns
* it will consider null values if there are NAN which is python default marker for null values
* you can use df.isnull() mehtods which is simply an alias of df.isna()

In [171]:
df_hr.isna()

satisfaction_level     le  number_project  average_montly_hours  \
0                   False  False           False                 False   
1                   False  False           False                 False   
2                   False  False           False                 False   
3                   False  False           False                 False   
4                   False  False           False                 False   
...                   ...    ...             ...                   ...   
14994               False  False           False                 False   
14995               False  False           False                 False   
14996               False  False           False                 False   
14997               False  False           False                 False   
14998               False  False           False                 False   

       time_spend_company  Work_accident   left  promotion_last_5years  \
0                   False          False  False                  False   
1                   False          False  False                  False   
2                   False          False  False                  False   
3                   False          False  False                  False   
4                   False          False  False                  False   
...                   ...            ...    ...                    ...   
14994               False          False  False                  False   
14995               False          False  False                  False   
14996               False          False  False                  False   
14997               False          False  False                  False   
14998               False          False  False                  False   

       Department  salary  
0           False   False  
1           False   False  
2           False   False  
3           False   False  
4           False   False  
...           ...     ...  
14994       False   False  
14995       False   False  
14996       False   False  
14997       False   False  
14998       False   False  

[14999 rows x 10 columns]

In [172]:
df_hr.isna().sum()

satisfaction_level       0
le                       0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
Department               0
salary                   0
dtype: int64

In [173]:
df_hr.notnull()

satisfaction_level    le  number_project  average_montly_hours  \
0                    True  True            True                  True   
1                    True  True            True                  True   
2                    True  True            True                  True   
3                    True  True            True                  True   
4                    True  True            True                  True   
...                   ...   ...             ...                   ...   
14994                True  True            True                  True   
14995                True  True            True                  True   
14996                True  True            True                  True   
14997                True  True            True                  True   
14998                True  True            True                  True   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                    True           True  True                   True   
1                    True           True  True                   True   
2                    True           True  True                   True   
3                    True           True  True                   True   
4                    True           True  True                   True   
...                   ...            ...   ...                    ...   
14994                True           True  True                   True   
14995                True           True  True                   True   
14996                True           True  True                   True   
14997                True           True  True                   True   
14998                True           True  True                   True   

       Department  salary  
0            True    True  
1            True    True  
2            True    True  
3            True    True  
4            True    True  
...           ...     ...  
14994        True    True  
14995        True    True  
14996        True    True  
14997        True    True  
14998        True    True  

[14999 rows x 10 columns]

In [178]:
df_hr.notna().sum(axis=0)

satisfaction_level       14999
le                       14999
number_project           14999
average_montly_hours     14999
time_spend_company       14999
Work_accident            14999
left                     14999
promotion_last_5years    14999
Department               14999
salary                   14999
dtype: int64

In [179]:
df_hr.dropna()

satisfaction_level    le  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

# fillna()

In [180]:
df_hr.fillna(5)

satisfaction_level    le  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

# Data Manipulaton
* Methods are:
  * replace()
    * replace values from DataFrame
    * df.replace({"old_val": "new_val", "old_val1": "new_val2"})
  * astype()
    * change DataFrame column types
    * df.astype({"col1": "new_type", "col2": "new_type2"})
  * drop()
  * sort_values()
  * groupby()
  * apply()

In [185]:
df_hr["left"].replace({1: "yes", 0: "no"}, inplace=True)
df_hr

satisfaction_level    le  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident left  promotion_last_5years  \
0                       3              0  yes                      0   
1                       6              0  yes                      0   
2                       4              0  yes                      0   
3                       5              0  yes                      0   
4                       3              0  yes                      0   
...                   ...            ...  ...                    ...   
14994                   3              0  yes                      0   
14995                   3              0  yes                      0   
14996                   3              0  yes                      0   
14997                   4              0  yes                      0   
14998                   3              0  yes                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

In [187]:
df_typecast = df_hr.astype({"left": "bool"})
df_typecast

satisfaction_level    le  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0  True                      0   
1                       6              0  True                      0   
2                       4              0  True                      0   
3                       5              0  True                      0   
4                       3              0  True                      0   
...                   ...            ...   ...                    ...   
14994                   3              0  True                      0   
14995                   3              0  True                      0   
14996                   3              0  True                      0   
14997                   4              0  True                      0   
14998                   3              0  True                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

In [188]:
df_drop = df_hr.drop("salary", axis=1)
df_drop

satisfaction_level    le  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident left  promotion_last_5years  \
0                       3              0  yes                      0   
1                       6              0  yes                      0   
2                       4              0  yes                      0   
3                       5              0  yes                      0   
4                       3              0  yes                      0   
...                   ...            ...  ...                    ...   
14994                   3              0  yes                      0   
14995                   3              0  yes                      0   
14996                   3              0  yes                      0   
14997                   4              0  yes                      0   
14998                   3              0  yes                      0   

      Department  
0          sales  
1          sales  
2          sales  
3          sales  
4          sales  
...          ...  
14994    support  
14995    support  
14996    support  
14997    support  
14998    support  

[14999 rows x 9 columns]

In [189]:
df_hr.sort_values(by="Work_accident", ascending=False)

satisfaction_level    le  number_project  average_montly_hours  \
8078                 0.77  0.91               5                   268   
3570                 0.50  0.60               3                   270   
1784                 0.43  0.51               2                   149   
6589                 0.64  0.64               3                   234   
11800                0.67  0.91               2                   245   
...                   ...   ...             ...                   ...   
5310                 0.63  0.87               5                   248   
5311                 0.86  0.75               5                   157   
5312                 0.80  0.79               5                   240   
5313                 0.55  0.58               5                   262   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident left  promotion_last_5years  \
8078                    3              1   no                      0   
3570                    3              1   no                      0   
1784                    3              1  yes                      0   
6589                    3              1   no                      0   
11800                   2              1   no                      1   
...                   ...            ...  ...                    ...   
5310                    2              0   no                      0   
5311                    4              0   no                      0   
5312                    2              0   no                      0   
5313                    3              0   no                      0   
14998                   3              0  yes                      0   

       Department  salary  
8078           IT  medium  
3570   management  medium  
1784   accounting  medium  
6589        sales    high  
11800       sales  medium  
...           ...     ...  
5310    technical     low  
5311      support     low  
5312      support     low  
5313      support  medium  
14998     support     low  

[14999 rows x 10 columns]

In [191]:
df_hr.sort_values(by="satisfaction_level", ascending=True)

satisfaction_level    le  number_project  average_montly_hours  \
14484                0.09  0.96               6                   296   
1873                 0.09  0.93               6                   279   
340                  0.09  0.89               7                   308   
1882                 0.09  0.89               6                   282   
332                  0.09  0.98               6                   271   
...                   ...   ...             ...                   ...   
12916                1.00  0.86               4                   245   
4553                 1.00  0.76               6                   246   
4555                 1.00  0.59               3                   168   
11122                1.00  0.77               5                   269   
4724                 1.00  0.73               3                   245   

       time_spend_company  Work_accident left  promotion_last_5years  \
14484                   4              0  yes                      0   
1873                    4              0  yes                      0   
340                     5              0  yes                      0   
1882                    4              0  yes                      0   
332                     4              0  yes                      0   
...                   ...            ...  ...                    ...   
12916                   4              0   no                      0   
4553                    3              1   no                      1   
4555                    3              0   no                      1   
11122                   3              0   no                      0   
4724                    2              1   no                      0   

        Department  salary  
14484    technical     low  
1873     technical     low  
340             hr  medium  
1882       support     low  
332          sales     low  
...            ...     ...  
12916        sales  medium  
4553       support  medium  
4555       support     low  
11122        sales     low  
4724   product_mng     low  

[14999 rows x 10 columns]

In [203]:
df_depart_left = df_hr.groupby(df_hr["Department"])
df_depart_left